File: backend/src/controllers/user.controller.ts
import { Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient();

export const userController = {
    // Get all users (admin only)
    async getAllUsers(req: Request, res: Response) {
        try {
            const users = await prisma.user.findMany({
                select: {
                    id: true,
                    userId: true,
                    name: true,
                    role: true,
                    walletBalance: true,
                    isActive: true,
                    createdAt: true,
                    updatedAt: true,
                },
                orderBy: {
                    createdAt: 'desc',
                },
            });

            res.json({ users });
        } catch (error) {
            console.error('Get all users error:', error);
            res.status(500).json({ error: 'Failed to fetch users' });
        }
    },

    // Get single user by ID (admin only)
    async getUserById(req: Request, res: Response) {
        try {
            const { id } = req.params;

            const user = await prisma.user.findUnique({
                where: { id },
                select: {
                    id: true,
                    userId: true,
                    name: true,
                    role: true,
                    walletBalance: true,
                    isActive: true,
                    createdAt: true,
                    updatedAt: true,
                },
            });

            if (!user) {
                return res.status(404).json({ error: 'User not found' });
            }

            res.json({ user });
        } catch (error) {
            console.error('Get user by ID error:', error);
            res.status(500).json({ error: 'Failed to fetch user' });
        }
    },

    // Create new user (admin only)
    async createUser(req: Request, res: Response) {
        try {
            const { userId, name, password, role, walletBalance, isActive } = req.body;

            // Validation
            if (!userId || !name || !password) {
                return res.status(400).json({ error: 'userId, name, and password are required' });
            }

            // Check if user already exists
            const existingUser = await prisma.user.findUnique({
                where: { userId },
            });

            if (existingUser) {
                return res.status(400).json({ error: 'User ID already exists' });
            }

            // Hash password
            const hashedPassword = await bcrypt.hash(password, 10);

            // Create user
            const user = await prisma.user.create({
                data: {
                    userId,
                    name,
                    password: hashedPassword,
                    role: role || 'CUSTOMER',
                    walletBalance: walletBalance || 0,
                    isActive: isActive !== undefined ? isActive : true,
                },
                select: {
                    id: true,
                    userId: true,
                    name: true,
                    role: true,
                    walletBalance: true,
                    isActive: true,
                    createdAt: true,
                    updatedAt: true,
                },
            });

            res.status(201).json({ user });
        } catch (error) {
            console.error('Create user error:', error);
            res.status(500).json({ error: 'Failed to create user' });
        }
    },

    // Update user (admin only)
    async updateUser(req: Request, res: Response) {
        try {
            const { id } = req.params;
            const { userId, name, password, role, walletBalance, isActive } = req.body;

            // Check if user exists
            const existingUser = await prisma.user.findUnique({
                where: { id },
            });

            if (!existingUser) {
                return res.status(404).json({ error: 'User not found' });
            }

            // If userId is being changed, check if new userId is available
            if (userId && userId !== existingUser.userId) {
                const userIdTaken = await prisma.user.findUnique({
                    where: { userId },
                });

                if (userIdTaken) {
                    return res.status(400).json({ error: 'User ID already exists' });
                }
            }

            // Prepare update data
            const updateData: any = {};
            if (userId) updateData.userId = userId;
            if (name) updateData.name = name;
            if (password) updateData.password = await bcrypt.hash(password, 10);
            if (role) updateData.role = role;
            if (walletBalance !== undefined) updateData.walletBalance = walletBalance;
            if (isActive !== undefined) updateData.isActive = isActive;

            // Update user
            const user = await prisma.user.update({
                where: { id },
                data: updateData,
                select: {
                    id: true,
                    userId: true,
                    name: true,
                    role: true,
                    walletBalance: true,
                    isActive: true,
                    createdAt: true,
                    updatedAt: true,
                },
            });

            res.json({ user });
        } catch (error) {
            console.error('Update user error:', error);
            res.status(500).json({ error: 'Failed to update user' });
        }
    },

    // Delete user (admin only)
    async deleteUser(req: Request, res: Response) {
        try {
            const { id } = req.params;

            // Check if user exists
            const existingUser = await prisma.user.findUnique({
                where: { id },
            });

            if (!existingUser) {
                return res.status(404).json({ error: 'User not found' });
            }

            // Delete user
            await prisma.user.delete({
                where: { id },
            });

            res.json({ message: 'User deleted successfully' });
        } catch (error) {
            console.error('Delete user error:', error);
            res.status(500).json({ error: 'Failed to delete user' });
        }
    },
};

File: server/backend/src/main/java/com/sanskriti/backend/controller/UserController.java
package com.sanskriti.backend.controller;

import com.sanskriti.backend.annotation.SuccessMessage;
import com.sanskriti.backend.dto.request.CreateUserRequest;
import com.sanskriti.backend.dto.request.UpdateUserRequest;
import com.sanskriti.backend.dto.response.UserResponse;
import com.sanskriti.backend.service.UserService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * UserController — replaces your Node.js user.routes.ts + user.controller.ts.
 *
 * Node.js routes (user.routes.ts):
 *   router.use(authenticateToken);       ← handled by SecurityConfig (all routes authenticated)
 *   router.use(requireAdmin);            ← @PreAuthorize("hasRole('ADMIN')") below
 *
 *   GET    /api/users        → getAllUsers()
 *   GET    /api/users/:id    → getUserById()
 *   POST   /api/users        → createUser()
 *   PUT    /api/users/:id    → updateUser()
 *   DELETE /api/users/:id    → deleteUser()
 *
 * 🎓 Key Concepts:
 *
 * @RestController
 *   = @Controller + @ResponseBody
 *   Every method returns JSON automatically (no res.json() needed).
 *
 * @RequestMapping("/api/users")
 *   Sets the base URL prefix for ALL methods in this class.
 *   Equivalent of: app.use('/api/users', userRoutes)
 *
 * @PreAuthorize("hasRole('ADMIN')")
 *   Applied at class level → ALL endpoints in this controller require ADMIN role.
 *   Replaces: router.use(authenticateToken); router.use(requireAdmin);
 *   If a non-admin calls these endpoints, Spring returns 403 automatically.
 *
 * @PathVariable
 *   Extracts :id from the URL path.
 *   Replaces: req.params.id
 *
 * @RequestBody
 *   Parses the JSON request body into the DTO.
 *   Replaces: req.body
 *
 * @Valid
 *   Triggers bean validation on the request body.
 *   If @NotBlank fails, GlobalExceptionHandler returns 400 automatically.
 *   Replaces: if (!userId || !name || !password) return res.status(400).json(...)
 *
 * @ResponseStatus(HttpStatus.CREATED)
 *   Returns HTTP 201 instead of 200 for the create endpoint.
 *   Replaces: res.status(201).json(...)
 *
 * @ResponseStatus(HttpStatus.NO_CONTENT)
 *   Returns HTTP 204 (no body) for the delete endpoint.
 *   Replaces: res.json({ message: 'User deleted successfully' })
 *   Note: 204 is more RESTful for DELETE — no body needed.
 *
 * Notice how THIN the controller is — no business logic here!
 * All logic lives in UserServiceImpl. This is the Spring Boot way.
 */
@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
@PreAuthorize("hasRole('ADMIN')") // All endpoints in this controller require ADMIN role
public class UserController {

    private final UserService userService;

    /**
     * GET /api/users
     * Returns all users (without passwords).
     * Replaces: async getAllUsers(_req, res) → res.json({ users })
     */
    @GetMapping
    @SuccessMessage("Users fetched successfully")
    public List<UserResponse> getAllUsers() {
        return userService.getAllUsers();
    }

    /**
     * GET /api/users/:id
     * Returns a single user by their UUID id.
     * Replaces: async getUserById(req, res) → res.json({ user })
     *
     * @PathVariable String id — extracts the {id} from the URL
     */
    @GetMapping("/{id}")
    @SuccessMessage("User fetched successfully")
    public UserResponse getUserById(@PathVariable String id) {
        return userService.getUserById(id);
    }

    /**
     * POST /api/users
     * Creates a new user (admin only).
     * Replaces: async createUser(req, res) → res.status(201).json({ user })
     */
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    @SuccessMessage("User created successfully")
    public UserResponse createUser(@Valid @RequestBody CreateUserRequest request) {
        return userService.createUser(request);
    }

    /**
     * PUT /api/users/:id
     * Updates an existing user (admin only).
     * Replaces: async updateUser(req, res) → res.json({ user })
     *
     * Note: We don't use @Valid here because all UpdateUserRequest fields are optional.
     * Only @Size on password needs checking — Spring validates it if password is provided.
     */
    @PutMapping("/{id}")
    @SuccessMessage("User updated successfully")
    public UserResponse updateUser(
            @PathVariable String id,
            @Valid @RequestBody UpdateUserRequest request
    ) {
        return userService.updateUser(id, request);
    }

    /**
     * DELETE /api/users/:id
     * Deletes a user (admin only).
     * Replaces: async deleteUser(req, res) → res.json({ message: 'User deleted' })
     *
     * Returns 204 No Content — RESTful standard for successful DELETE.
     */
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteUser(@PathVariable String id) {
        userService.deleteUser(id);
    }
}
